torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1373,126 +1374,122 @@ void ONNXUpdateTypeFromTensor(
+  } else if (PyTuple_Check(output_obj)) {
+    size_t tuple_len = PyTuple_GET_SIZE(output_obj);
+    for (size_t i = 0; i < tuple_len; ++i) {
+      outputs_index = ONNXAssignOutputShape(
+          graph,
+          outputs_index,
+          PyTuple_GET_ITEM(output_obj, i),
+          onnx_shape_inference);
+    }
+  } else if (PyList_Check(output_obj)) {
+    size_t list_len = PyList_GET_SIZE(output_obj);
+    if (HasSequenceTypeOutput(graph->outputs().at(outputs_index)->node())) {
+      if (list_len > 0) {
+        auto list_elem = PyList_GET_ITEM(output_obj, 0);
+        TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+        auto& var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+        for (size_t i = 1; i < list_len; ++i) {
+          list_elem = PyList_GET_ITEM(output_obj, i);
+          TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+          auto& new_var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+          TORCH_CHECK(
+              var.scalar_type() == new_var.scalar_type(),
+              "Unsupported sequence type in model outputs. ONNX supports sequences of elements of the same data type.");
+        }
+        auto elem_type = graph->outputs()
+                             .at(outputs_index)
+                             ->type()
+                             ->castRaw<ListType>()
+                             ->getElementType()
+                             ->cast<TensorType>();
+        elem_type = elem_type->withScalarType(var.scalar_type());
+        graph->outputs()
+            .at(outputs_index)
+            ->setType(MergeInferredType(
+                graph->outputs().at(outputs_index)->type(),
+                ListType::create(elem_type)));
+      } else {
+        graph->outputs()
+            .at(outputs_index)
+            ->setType(graph->outputs().at(outputs_index)->type());
+      }
+      outputs_index++;
+    } else {
+      // When torch output is a list type, but ONNX node is not a
+      // sequence type. Like prim::ListConstruct
+      for (size_t i = 0; i < list_len; ++i) {
+        outputs_index = ONNXAssignOutputShape(
+            graph,
+            outputs_index,
+            PyList_GET_ITEM(output_obj, i),
+            onnx_shape_inference);
+      }
+    }
+  } else if (PyDict_Check(output_obj)) {
+    // Support for dict data type is limited to fixed size dictionaries in
+    // ONNX.
+    // Dictionary values are unrolled and keys are not preserved.
+    auto unrolled_dict =
+        py::reinterpret_borrow<py::list>(PyDict_Items(output_obj));
+    TORCH_INTERNAL_ASSERT(PyList_Check(unrolled_dict.ptr()));
+    for (size_t i = 0; i < unrolled_dict.size(); ++i) {
+      outputs_index = ONNXAssignOutputShape(
+          graph,
+          outputs_index,
+          PyList_GET_ITEM(unrolled_dict.ptr(), i),
+          onnx_shape_inference);
+    }
+  } else if (THPUtils_checkString(output_obj)) {
+    // Ignore string, since they are not supported as output in ONNX.
```

Comment:
Should we through a warning or error in this case?

Reply:
Similar to above, i think it's ok to not throw warnings here, since use of strings and dictionaries are already warned elsewhere.